torch/csrc/jit/mobile/versioned_operators.cpp

```diff
@@ -0,0 +1,185 @@
+// names and values are a container with version numbers and their metadata. The
+// incoming operator with version is searched in the table to find the match.
+// Note that the size of op_version_table is small because it only contains
+// operators that have more than one versions. For most of the operators, they
+// fall through to the existing implementation with default version 0.
+
+namespace torch {
+namespace jit {
+namespace mobile {
+namespace {
+
+// The version info is limited to combinations of two situations:
+//   * Route to another op name
+//   * Push additional default values
+// It’s fine to update the content of version info, to handle more situations in
+// fugure. Since it’s implementation details, it should not affect
+// compatibility.
+struct VersionInfo {
+  VersionInfo(
+      int64_t version,
+      const c10::OperatorName& name,
+      c10::optional<IValue> default_val)
+      : version(version), name(name), default_val(default_val) {}
+  int64_t version;
+  c10::OperatorName name;
+  c10::optional<IValue> default_val;
+};
+
+struct VersionInfoContainer {
+  VersionInfoContainer(const std::vector<VersionInfo>& version_vec)
+      : version_vec(version_vec) {}
+  int64_t min_version() const {
+    return version_vec.begin()->version;
+  }
+
+  int64_t max_version() const {
+    return version_vec.rbegin()->version + 1;
+  }
+
+  std::vector<VersionInfo> version_vec;
+};
+
+VersionInfo getVersioninfo(
+    const VersionInfoContainer& container,
+    const c10::OperatorName& name,
+    int64_t op_version) {
+  // Backward compatibility breakage
+  TORCH_CHECK(
+      op_version >= container.min_version(),
+      "The version number of ",
+      c10::toString(name),
+      ", ",
+      op_version,
+      ", is smaller than the minimum version number, ",
+      container.min_version(),
+      ", supported in this runtime. The model file is too old. ",
+      "Update the model with most recent code.");
+
+  // Forward compatibility breakage
+  TORCH_CHECK(
+      op_version <= container.max_version(),
+      "The version number of ",
+      c10::toString(name),
+      ", ",
+      op_version,
+      ", is larger than the maximum version number, ",
+      container.max_version(),
+      ", supported in this runtime. Update the runtime to be compatible",
+      " to this model.");
+
+  for (const auto& info : container.version_vec) {
+    if (info.version == op_version) {
+      return info;
+    }
+  }
+
+  if (op_version == container.max_version() + 1) {
+    // Current version
+    return VersionInfo(op_version, name, c10::nullopt);
+  }
+
+  // The situation where op_version is in [min_version, max_version + 1], but
+  // not in the container, which should not happen.
+  TORCH_CHECK(
+      true,
+      "The version number of ",
+      c10::toString(name),
+      ", ",
+      op_version,
+      " is not compatible in this runtime. ");
+
+  return VersionInfo(
```

Comment:
I think you can eliminate this return once the above check always starts throwing.